### PR TITLE
Improve Micro::State

### PR DESCRIPTION
**Pull request by linqueta**

### CODE REVIEW

File: lib/micro/case.rb [hunk 1]
```diff
@@ -5,14 +5,14 @@
...
-    require 'micro/case/flow'
-    require 'micro/case/safe/flow'
+    require_relative 'case/version'
+    require_relative 'case/result'
+    require_relative 'case/error'
+    require_relative 'case/safe'
+    require_relative 'case/strict'
+    require_relative 'case/flow/reducer'
+    require_relative 'case/flow'
+    require_relative 'case/safe/flow'
```
Comment (serradura) [inline lib/micro/case.rb]: @linqueta, thanks for your contribution. But, I think that is an unnecessary change because the RubyGems resolves the gems' paths, and I don't see this kind of approach in other ones (.e.g. https://github.com/rails/rails/blob/master/activesupport/lib/active_support.rb#L28-L30 ).

File: lib/micro/case.rb [hunk 2]
```diff
@@ -37,9 +37,7 @@ def self.call(options = {})
     end
 
     def self.__new__(result, arg)
-      instance = new(arg)
-      instance.__set_result__(result)
-      instance
+      new(arg).tap { |instance| instance.__set_result__(result) }
```
Comment (serradura) [inline lib/micro/case.rb]: @linqueta Nice refactoring! But it adds a little overhead (please, check out the below example), and yes, I know that is a micro-optimization, but one of the project's goals is: `Be fast and optimized`. So I think that this kind of change would be unnecessary.

Following this discussion, what do you think about the refactoring to replace the usage of the [method `tap`](https://github.com/serradura/u-case/blob/master/lib/micro/case/result.rb#L44-L52) to increase the `u-case` performance?

```ruby
require 'bundler/inline'

gemfile do
  source 'https://rubygems.org'

  gem 'benchmark-ips', '~> 2.7', '>= 2.7.2'
end

require 'benchmark/ips'

Result = Struct.new(:value)

class Foo
  def self.__new__a(value, result)
    new(value).tap { |instance| instance.__set_result__(result) }
  end

  def self.__new__b(value, result)
    instance = new(value)
    instance.__set_result__(result)
    instance
  end

  def initialize(value)
    @value = value 
  end

  def __set_result__(result)
    @result = result
  end
end

result = Result.new(nil)

Benchmark.ips do |x|
  x.config(:time => 5, :warmup => 2)

  x.time = 5
  x.warmup = 2
  
  x.report('with tap') { Foo.__new__a(1, result) }
  
  x.report('without tap') { Foo.__new__b(2, result) }

  x.compare!
end
```

**Comment by linqueta** [comment]
Wow, nice @serradura . I'm grateful about the answers and I agree with all that you said.
